rfd/00XX-proxy-tunnel-service.md

```diff
@@ -0,0 +1,139 @@
+message Frame {
+   oneof Message {
+       DialRequest DialRequest = 1;
+       Data Data = 2;
+   }
+}
+
+// DialRequest contains details for connecting to a node.
+message DialRequest {
+   // NodeID is the {UUID}.{ClusterName} of the node to connect to.
+   string NodeID = 1;
+   // TunnelType is the type of service being accessed. This differentiates agents that
+   // create multiple reverse tunnels for different services.
+   string TunnelType = 2 [ (gogoproto.casttype) = "github.com/gravitational/teleport/api/types.TunnelType" ];
+   // Source is the original source address of the client.
+   Addr Source = 3;
+   // Destination is the destination address to connect to over the reverse tunnel.
+   Addr Destination = 4;
+}
+
+message Addr {
+   // Network is the name of the network transport.
+   string Network = 1;
+   // String is the string form of the address.
+   string String = 2;
+}
+
+// Data contains the raw bytes of a connection.
+message Data { bytes Bytes = 1; }
+```
+
+### How it works
+
+The following diagram shows a user connecting to a proxy, the user-proxy, and trying to reach a node connected to a different proxy, the node-proxy. Using the DialNode rpc, the user-proxy can create a bidirectional stream to the node through the node-proxy.
+```
+┌────┐                         ┌──────────┐
+|user|──────connection────────>|user-proxy|
+└────┘                         └────╥─────┘
+                                    ║
+                                grpc stream
+                                    ║
+┌────┐                         ┌────╨─────┐
+|node|─────reverse-tunnel─────>|node-proxy|
+└────┘                         └──────────┘
+```
+
+A call to the DialNode rpc will send an initial frame containing a `DialRequest`. All subsequent frames should contain `Data` messages. An appropriate error will be returned if the dial request fails for some reason.
+
+To avoid duplicate work the user-proxy will handle all typical proxy side logic like authorization and session recording, while the node-proxy will forward the connection directly to the node.
+
+The DialNode rpc will be wrapped with a client library to return a net.Conn when called. This abstraction allows teleport to treat any underlying transport the same, whether it be a direct dial to the node, a reverse tunnel connected to the user-proxy, or a connection over the DialNode rpc.
+
+```go
+type NodeTunnelClient interface {
+    DialNode(
+        ctx context.Context,
+        proxyAddress string,
+        src net.Addr,
+        dst net.Addr,
+        serverID string,
+        tunnelType types.TunnelType,
+    ) (net.Conn, error)
+}
+```
+
+### Security
+The api will use mTLS to ensure that only other proxies are able to connect. This is done by checking certificates for the build-in role “Proxy”. This will prevent users from connecting to the service directly without going through the user-proxy logic of authorization and session recording.
```

Comment:
Nitpick, I think it's built-in. 
```suggestion
The api will use mTLS to ensure that only other proxies are able to connect. This is done by checking certificates for the built-in role “Proxy”. This will prevent users from connecting to the service directly without going through the user-proxy logic of authorization and session recording.
```